diff --git a/PyInstaller/building/build_main.py b/PyInstaller/building/build_main.py
index ef48170b..13090178 100644
--- a/PyInstaller/building/build_main.py
+++ b/PyInstaller/building/build_main.py
@@ -585,6 +585,8 @@ class Analysis(Target):
             new_toc = TOC()
             for name, path, typecode in self.pure:
                 assert typecode == 'PYMODULE'
+                if 'pyinstaller.exe' in path:
+                    continue
                 # Transform a python module name into a file name.
                 name = name.replace('.', os.sep)
                 # Special case: modules have an implied filename to add.
