Index: orbit-python/corba-object.c
===================================================================
RCS file: /cvs/gnome/orbit-python/orbit-python/corba-object.c,v
retrieving revision 1.2
diff -u -p -r1.2 corba-object.c
--- orbit-python/corba-object.c	31 Jan 2002 14:34:01 -0000	1.2
+++ orbit-python/corba-object.c	26 Oct 2002 15:05:18 -0000
@@ -121,7 +121,7 @@ CORBA_Object_to_PyObject_with_type (CORB
 	}
 
 	inst_glue = g_new0 (CORBA_PyInstance_Glue, 1);
-	inst_glue->obj        = object;
+	inst_glue->obj        = CORBA_Object_duplicate (object, &inst_glue->ev);
 	inst_glue->class_glue = class_glue;
 	inst_glue->repo_id    = repo_id;
 	inst_glue->orb        = orb;
Index: orbit-python/corba-orb.c
===================================================================
RCS file: /cvs/gnome/orbit-python/orbit-python/corba-orb.c,v
retrieving revision 1.2
diff -u -p -r1.2 corba-orb.c
--- orbit-python/corba-orb.c	31 Jan 2002 14:34:01 -0000	1.2
+++ orbit-python/corba-orb.c	26 Oct 2002 15:05:19 -0000
@@ -114,7 +114,8 @@ CORBA_ORB_PyObject__new (CORBA_ORB orb)
 
 	CORBA_exception_init (&self->ev);
 
-	self->obj = orb;
+	self->obj = (CORBA_ORB) CORBA_Object_duplicate ((CORBA_Object)orb, &self->ev);
+	//self->obj = orb;
 	g_hash_table_insert (orb_objects, orb, self);
 	
 	return self;
